### DISCUSSION TITLE
jsdom vs happy-dom

### DISCUSSION THREAD

**Discussion by cfife-btig** [vitest-dev/vitest#1607] [General]
Hi there, I'm not sure if this is the place to ask this, but I'm a little confused by how there seems to be two different engines for rendering the dom, happy-dom and jsdom. Could someone share why some of the vitest examples seem to have both and roughly what their differences are?

**Comment by sheremet-va** [comment]
happy-dom is a faster alternative, but it doesn't have some Web APIs. jsdom is more mature package.

**Reply by espetro** [reply]
Which are those _some Web APIs_?? I keep seeing this mention to `happy-dom` but cannot find the listing of missing Web APIs.

**Reply by belgattitude** [reply]
It can be found in the sources:

- jsdom: https://github.com/jsdom/jsdom/tree/master/lib/jsdom/living
- happy-dom: https://github.com/capricorn86/happy-dom/tree/master/packages/happy-dom/src

On most project happy-dom works fantastic, I'd start with it. Open an issue whenever it lacks something, its author is very reactive in my experience. For more complex projects, jsdom is amazing but a bit slower.

**Reply by knownasilya** [reply]
happy-dom also has more bugs. I tried it but had to move back to jsdom for now.

**Reply by vhscom** [reply]
@knownasilya Could you share what specific bug or bugs you experienced that where the deal breaker?

**Reply by knownasilya** [reply]
https://github.com/capricorn86/happy-dom/issues/1079
https://github.com/capricorn86/happy-dom/issues/1181

Plus others I'm sure, but I moved back to jsdom after those two. Seems less spread out maintenance wise (one main maintainer).

**Reply by ghost** [reply]
I had this weird bug with happy-dom while doing `el.replaceChild(newNode, el.childNodes[index])`. Even though `el` had childNodes, it was throwing error that `Node to be inserted before is not the child of element`.

With jsdom, it was not erroring out.

**Reply by samal-rasmussen** [reply]
> [capricorn86/happy-dom#1079](https://github.com/capricorn86/happy-dom/issues/1079)
> [capricorn86/happy-dom#1181](https://github.com/capricorn86/happy-dom/issues/1181)

Both of those issues are resolved and a lot of other feature implemented since. Lots of activity recently and consistently. The happy-dom dev seems super effective.

**Reply by ItamarGronich** [reply]
> For more complex projects, jsdom is amazing but a bit slower.

@belgattitude jsdom can't submit forms....
https://github.com/jsdom/jsdom/issues/1937

**Comment by mindplay-dk** [comment]
I think a lot of us wish for a more lightweight DOM mock for quick unit tests.

I was quite excited about [`undom`](https://github.com/developit/undom) but ran into various problems using that as well.

I'm afraid, at this time, `jsdom` remain the devil you know.

You wouldn't think it'd be all that difficult to replicate just the most essential DOM features in a more lightweight way - just a simpler package, with a smaller scope, without mocking the entire browser, but still accurate to the specs.

But I dunno... the DOM specs are awful complex. 🤔
